src/dataflow-types/src/types.rs

```diff
@@ -351,52 +351,110 @@ impl SourceDataEncoding {
-
-                        // Add the key columns as a key.
-                        let key_indices = (0..key_desc.arity()).collect();
-                        let key_desc = key_desc.with_key(key_indices);
-
-                        // Rename key columns to "keyN" if the encoding is not Avro.
-                        match key {
-                            DataEncoding::Avro(_) => key_desc,
-                            _ => {
-                                let names =
-                                    (0..key_desc.arity()).map(|i| Some(format!("key{}", i)));
-                                key_desc.with_names(names)
-                            }
-                        }
-                    };
-                    let key_schema = if let DataEncoding::Avro(AvroEncoding { schema, .. }) = key {
-                        Some(&**schema)
-                    } else {
-                        None
-                    };
-                    value.desc(envelope, key_desc, key_schema)
-                } else {
-                    value.desc(envelope, RelationDesc::empty(), None)
-                };
-                desc
-            }
+            SourceDataEncoding::KeyValue { key, value } => match key_envelope {
+                KeyEnvelope::None => value.desc(envelope, RelationDesc::empty(), None),
+                KeyEnvelope::Flattened => insert_flattened_key(key, value, envelope, false),
+                KeyEnvelope::LegacyUpsert => insert_flattened_key(key, value, envelope, true),
+                KeyEnvelope::Named(key_name) => insert_named_key(key_name, key, value, envelope),
+            },
         }
     }
 }
 
+/// modify the desc to include the key using th appropriate name
+fn insert_named_key(
+    key_name: &str,
+    key: &DataEncoding,
+    value: &DataEncoding,
+    envelope: &SourceEnvelope,
+) -> Result<RelationDesc, anyhow::Error> {
+    let key_desc = {
+        let key_desc = key.desc(&SourceEnvelope::None, RelationDesc::empty(), None)?;
+
+        // It doesn't make sense for the key to have keys.
+        assert!(key_desc.typ().keys.is_empty());
+
+        // if the key has multiple objects, nest them as a record inside of a single name
+        if key_desc.arity() > 1 {
+            let key_type = key_desc.typ();
+            let key_as_record = RelationType::new(vec![ColumnType {
+                nullable: false,
+                scalar_type: ScalarType::Record {
+                    fields: key_desc
+                        .iter_names()
+                        .enumerate()
+                        .map(|(i, name)| {
+                            name.map(|n| n.clone())
+                                .unwrap_or_else(|| format!("key{}", i).into())
+                        })
+                        .zip(key_type.column_types.iter())
+                        .map(|(name, typ)| (name, typ.clone()))
+                        .collect(),
+                    // TODO: what do these do
```

Comment:
`custom_oid` is probably irrelevant for your purpose here. `custom_name` will cause the type to be emitted in the Avro schema with a particular name if someone makes a sink based off of thiss